Adds galbot stack cube tasks, with left_arm_gripper and right_arm_suction, using RMPFlow controller

rebeccazhang0707 opened the pull request:
# Description

<!--
Thank you for your interest in sending a pull request. Please make sure to check the contribution guidelines.

Link: https://isaac-sim.github.io/IsaacLab/main/source/refs/contributing.html
-->

Adds galbot_stack_cube tasks and mimic tasks, using RMPFlow controller:
- add galbot robot config and .usd asset
- add RMPFlowAction and RMPFlowActionCfg
- add motion_policy_configs and .urdf for both 'galbot_left_arm_gripper' and 'galbot_right_arm_suction'
- add new task: galbot stack_rmp_rel_env_cfg
- add new mimic task: galbot_stack_rmp_abs/rel_mimic_env
- add mdp.observations/terminations/events for galbot: support gripper_state checking for both suction_cup and parallel_gripper, get obs_in_base_frame 
- add gripper_configs (gripper_joint_names, gripper_open_val, gripper_threshold) in galbot/franka tasks: to make mdp functions universal to varied robots
- fix a bug (eef_name) in franka_stack_ik_rel_mimic_env.py
- add new device_name in se3_spacemouse.py

Notes: This PR relies on PR (https://github.com/isaac-sim/IsaacLab/pull/3174) for surface gripper support in manager-based workflow.

You can test the whole gr00t-mimic workflow by:
1. record demos:
`./isaaclab.sh -p scripts/tools/record_demos.py --task Isaac-Stack-Cube-Galbot-Left-Arm-Gripper-RmpFlow-v0 --teleop_device spacemouse --num_demos 1 --device cpu --dataset_file datasets/recorded_demos_galbot_suction_rel.hdf5`
2. replay demos:
`./isaaclab.sh -p scripts/tools/replay_demos.py --task Isaac-Stack-Cube-Galbot-Left-Arm-Gripper-RmpFlow-v0 --num_envs 1 --device cpu --dataset_file datasets/recorded_demos_galbot_suction_rel.hdf5`
3. annotate demos:
`./isaaclab.sh -p scripts/imitation_learning/isaaclab_mimic/annotate_demos.py --task Isaac-Stack-Cube-Galbot-Left-Arm-Gripper-RmpFlow-Rel-Mimic-v0 --auto --device cpu --input_file datasets/recorded_demos_galbot_suction_rel.hdf5 --output_file datasets/annotated_demos_galbot_suction_rel.hdf5`
4. generate dataset:
`./isaaclab.sh -p scripts/imitation_learning/isaaclab_mimic/generate_dataset.py --task Isaac-Stack-Cube-Galbot-Left-Arm-Gripper-RmpFlow-Rel-Mimic-v0 --num_envs 16 --device cpu --generation_num_trials 10 --input_file datasets/annotated_demos_galbot_suction_rel.hdf5 --output_file datasets/generated_demos_galbot_suction_rel.hdf5`

<!-- As a practice, it is recommended to open an issue to have discussions on the proposed pull request.
This makes it easier for the community to keep track of what is being developed or added, and if a given feature
is demanded by more than one party. -->

## Type of change

<!-- As you go through the list, delete the ones that are not applicable. -->

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## Screenshot
<img width="1428" height="281" alt="environments_galbot" src="https://github.com/user-attachments/assets/b603bc60-7ff3-44e8-bf41-ab8126a753ec" />


## Checklist

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

<!--
As you go through the checklist above, you can mark something as done by putting an x character in it

For example,
- [x] I have done this task
- [ ] I have not done this task
-->

rebeccazhang0707 committed:
Add surface gripper in manager-based workflow
- add surfacegripper actions
- add new task envs with UR10 + suction cups

rebeccazhang0707 committed:
pre-commit format check

rebeccazhang0707 committed:
adapt to official Sim usd assets with gripper variants, and remove redundant parallel_gripper task

rebeccazhang0707 committed:
fix bugs in SurfaceGripperAction apply_actions(), dimension alignment

rebeccazhang0707 committed:
add galbot_stack_cube tasks and mimic tasks:
- add galbot robot config
- add RMPFlowAction and RMPFlowActionCfg
- add motion_policy_configs for both galbot_left_arm_gripper and galbot_right_arm_suction
- add galbot stack_rmp_rel_env_cfg
- add galbot_stack_rmp_abs/rel_mimic_env
- add mdp.observations/terminations/events for galbot: support both suction_cup and parallel_gripper, get obs_in_base_frame
- fix a bug (eef_name) in franka_stack_ik_rel_mimic_env.py
- add new device_name in se3_spacemouse.py

rebeccazhang0707 committed:
- use "reset" events to modify the robot initial_states, and convert the RobotCfg back to default
- move surface_gripper_actions out of binary_joint_actions to a standalone file

rebeccazhang0707 committed:
Merge branch 'reb/surface_gripper' into reb/galbot_stack_cube

rebeccazhang0707 committed:
after ./isaaclab.sh -f

rebeccazhang0707 committed:
add AbsBinaryJointPositionAction

rebeccazhang0707 committed:
pre-commit format check

rebeccazhang0707 committed:
fixes surface_gripper operations in get_state() and reset_to() of InteractiveScene class

rebeccazhang0707 committed:
smaller pos_sensitivity for keyboard teleop

rebeccazhang0707 committed:
Merge branch 'reb/surface_gripper' into reb/galbot_stack_cube

rebeccazhang0707 committed:
decrease the cube_scale to default, otherwise, the cube_stacked() terminiation function cannot be reached

rebeccazhang0707 committed:
Merge branch 'reb/surface_gripper' into reb/galbot_stack_cube

rebeccazhang0707 committed:
Merge branch 'main' into reb/surface_gripper

Signed-off-by: rebeccazhang0707 <168459200+rebeccazhang0707@users.noreply.github.com>

rebeccazhang0707 committed:
Revert "decrease the cube_scale to default, otherwise, the cube_stacked() terminiation function cannot be reached"

This reverts commit 6edd4028bbda25af97cc6e4278978cb9004ee68d.

rebeccazhang0707 committed:
fix bug of surface_gripper checking in obs and terminations:
- env.scene.surface_grippers is a dict

rebeccazhang0707 committed:
remove randomize_galbot_joint_by_gaussion_offset event, which is not used

rebeccazhang0707 committed:
refactor SurfaceGripperBinaryAction class

rebeccazhang0707 committed:
revise the return_key options in get cube_poses and ee_frame_pose

rebeccazhang0707 committed:
revert the original UR10_CFG

kellyguo11 committed:
Merge branch 'main' into reb/surface_gripper

rebeccazhang0707 committed:
- UR10_LONG/SHORT_SUCTION_CFG inherits UR10_CFG and overrides some variants
- decrease cube_scale to original, otherwise, cube_stacked termination func cannot be reachced

rebeccazhang0707 committed:
update original surface_gripper scripts

rebeccazhang0707 committed:
format check

rebeccazhang0707 committed:
update docs

kellyguo11 committed:
Merge branch 'main' into reb/galbot_stack_cube

kellyguo11 commented on source/isaaclab/isaaclab/controllers/config/motion_policy_configs/galbot_one_charlie/galbot_one_charlie.urdf:
```diff
@@ -0,0 +1,1157 @@
+<?xml version="1.0" ?>
```
can we upload this to https://isaac-dev.ov.nvidia.com/omni/web3/omniverse://isaac-dev.ov.nvidia.com/Isaac/IsaacLab/Controllers/ similar to how the curobo urdfs are stored? We can ask Sheikh to push them publicly once we are ready.

rebeccazhang0707 replied:
Done, but I found an existing [lula_franka_gen.urdf](https://github.com/isaac-sim/IsaacLab/blob/main/source/isaaclab/isaaclab/controllers/config/data/lula_franka_gen.urdf) for Rmflow controller in Lab's source code.

kellyguo11 commented on source/isaaclab/isaaclab/envs/mdp/actions/__init__.py:
```diff
@@ -10,3 +10,4 @@
 from .joint_actions import *
 from .joint_actions_to_limits import *
 from .non_holonomic_actions import *
+from .surface_gripper_actions import *
```
does this cause any performance regressions in startup time? I believe we only enable the surface gripper extension if it's needed to prevent the extra dependencies from starting up if we don't need them. does adding the import here always trigger the extension load up?

rebeccazhang0707 replied:
Yes, but I found the isaacsim.robot.surface_gripper-3.2.6 extension is already triggered before my commits. Any tools for me to compare the startup time? thanks.
<img width="793" height="544" alt="image" src="https://github.com/user-attachments/assets/738db4d9-50fe-4761-8d04-06f89b823f4a" />

kellyguo11 commented on source/isaaclab/isaaclab/envs/mdp/actions/actions_cfg.py:
```diff
@@ -9,7 +9,14 @@
+    joint_actions,
+    joint_actions_to_limits,
+    non_holonomic_actions,
+    surface_gripper_actions,
```
not sure if this will also trigger extension startup

kellyguo11 commented on source/isaaclab/isaaclab/envs/mdp/actions/actions_cfg.py:
```diff
@@ -181,6 +188,41 @@ class BinaryJointVelocityActionCfg(BinaryJointActionCfg):
     class_type: type[ActionTerm] = binary_joint_actions.BinaryJointVelocityAction
 
 
+@configclass
```
@ooctipus can you please help review the action term related changes?

kellyguo11 commented on source/isaaclab_assets/isaaclab_assets/robots/galbot.py:
```diff
@@ -0,0 +1,118 @@
+
+GALBOT_ONE_CHARLIE_CFG = ArticulationCfg(
+    spawn=sim_utils.UsdFileCfg(
+        usd_path=f"{ISAACLAB_ASSETS_DATA_DIR}/Robots/Galbot/galbot_one_charlie.usd",
```
we should upload assets to the Isaac/Robots folder - https://isaac-dev.ov.nvidia.com/omni/web3/omniverse://isaac-dev.ov.nvidia.com/Isaac/Robots/

rebeccazhang0707 replied:
@kellyguo11  Uploaded the assets to nucleus server.

kellyguo11 commented on source/isaaclab_assets/isaaclab_assets/robots/galbot.py:
```diff
@@ -0,0 +1,118 @@
+            joint_names_expr=["head_joint.*"],
+            velocity_limit_sim=1.50,
+            effort_limit_sim=4.0,
+            stiffness=17453293764608.0,
```
that seems like an extremely high value for stiffness. are we sure these are correct values that can also work for the real robot?

rebeccazhang0707 replied:
hi @kellyguo11 thanks for the coments, these KPs, KDs are from our customer's USD, and I think they just tuned these params with **Gain Tuner** in Isaac Sim. I do not think this works for real robot. Do you think this matters? Since Sim2Real SystemID is still a research topic, I think we cannot realize PDs to work both in Sim and Real. But I can set here all stiffness/damping/effort_limit to **None** to use default setup in the robot usd. Do you think it makes sense?

kellyguo11 replied:
I think if we can get it working with more realistic values, then it'd be good to keep those so that we won't be too far off from what real robots would expect.

rebeccazhang0707 replied:
@kellyguo11 , all gains and effort/velocity_limits are set to None here, kept the original settings in USD.

kellyguo11 commented on source/isaaclab_mimic/isaaclab_mimic/envs/__init__.py:
```diff
@@ -14,6 +14,13 @@
 from .franka_stack_ik_rel_mimic_env_cfg import FrankaCubeStackIKRelMimicEnvCfg
 from .franka_stack_ik_rel_visuomotor_cosmos_mimic_env_cfg import FrankaCubeStackIKRelVisuomotorCosmosMimicEnvCfg
 from .franka_stack_ik_rel_visuomotor_mimic_env_cfg import FrankaCubeStackIKRelVisuomotorMimicEnvCfg
+from .galbot_stack_rmp_abs_mimic_env import RmpFlowGalbotCubeStackAbsMimicEnv
```
@peterd-NV would be great if you could help review the mimic environment changes

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/stack/config/galbot/stack_joint_pos_env_cfg.py:
```diff
@@ -0,0 +1,318 @@
+            # 7-12 'head_joint2', 'left_arm_joint2', 'right_arm_joint2', 'left_arm_joint3', 'right_arm_joint3', 'left_arm_joint4',
+            # 13-18 'right_arm_joint4', 'left_arm_joint5', 'right_arm_joint5', 'left_arm_joint6', 'right_arm_joint6', 'left_arm_joint7',
+            # 19-22 'right_arm_joint7', 'left_gripper_left_joint', 'left_gripper_right_joint', 'right_suction_cup_joint1']
+            "default_pose": [
```
is there a reason why we can't use the init pose defined in the asset cfg?

rebeccazhang0707 replied:
Thanks Kelly, fixed this to use that in asset cfg.

kellyguo11 commented on source/isaaclab_tasks/isaaclab_tasks/manager_based/manipulation/stack/config/galbot/__init__.py:
```diff
@@ -0,0 +1,74 @@
+# SPDX-License-Identifier: BSD-3-Clause
+
+
+import gymnasium as gym
```
@ooctipus @peterd-NV would be good if you can also help review the stack environment updates

rebeccazhang0707 committed:
fixs  ArticulaitonCfg to use default gains setup from USD in galbot.py
- updates reset events to use init pos from Asset Cfg
- fixs a bug of head_camera pose

rebeccazhang0707 committed:
update the usd_path with variants

rebeccazhang0707 committed:
move the rmpflow configs to cloud

rebeccazhang0707 committed:
adds to environments in docs

rebeccazhang0707 committed:
Merge branch 'reb/surface_gripper' into reb/galbot_stack_cube

rebeccazhang0707 committed:
add retrieve_file_path for RmpFlowAssets Cfg files

kellyguo11 committed:
Merge branch 'main' into reb/surface_gripper

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 commented on source/isaaclab/isaaclab/controllers/config/rmp_flow.py:
```diff
@@ -9,6 +9,13 @@
+# FIXME: temp uploaded to nucleus folder, remove this line in public release and use above ISAACLAB_NUCLEUS_DIR
+ISAACLAB_NUCLEUS_DIR = "https://isaac-dev.ov.nvidia.com/omni/web3/omniverse://isaac-dev.ov.nvidia.com/Isaac/IsaacLab"
+
+ISAACLAB_NUCLEUS_RMPFLOW_DIR = os.path.join(ISAACLAB_NUCLEUS_DIR, "Controllers", "RmpFlowAssets")
```
once we are ready with this folder, we can ping Sheikh to push these changes publicly

rebeccazhang0707 committed:
update envs in comprehensive list

rebeccazhang0707 committed:
Merge branch 'reb/surface_gripper' of https://github.com/rebeccazhang0707/IsaacLab into reb/surface_gripper

rebeccazhang0707 committed:
Merge branch 'reb/surface_gripper' into reb/galbot_stack_cube

rebeccazhang0707 committed:
add envs to Comprehensive List

kellyguo11 reviewed (commented):
should be good to go once assets are available on S3.

rebeccazhang0707 committed:
update usd path to S3 bucket and fix typo in environments

rebeccazhang0707 committed:
update the rmpflow asset config path to S3

kellyguo11 committed:
Merge branch 'main' into reb/galbot_stack_cube

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Update CHANGELOG.rst

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Update extension.toml

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 commented on docs/source/overview/environments.rst:
```diff
@@ -3,7 +3,8 @@
 
-The following lists comprises of all the RL or IL tasks implementations that are available in Isaac Lab.
+The following lists comprises of all the RL and IL tasks implementations that are available in Isaac Lab.
+
```
```suggestion
```

kellyguo11 committed:
Update docs/source/overview/environments.rst

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Update CHANGELOG.rst

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Update extension.toml

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Merge branch 'main' into reb/galbot_stack_cube

kellyguo11 committed:
Merge branch 'main' into reb/galbot_stack_cube

kellyguo11 committed:
Merge branch 'main' into reb/galbot_stack_cube

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Update extension.toml

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Merge branch 'main' into reb/galbot_stack_cube

Signed-off-by: Kelly Guo <kellyg@nvidia.com>

kellyguo11 committed:
Update extension.toml

Signed-off-by: Kelly Guo <kellyg@nvidia.com>
